File: apps/mobile/android/app/src/main/kotlin/dev/anvil/LifecycleAdapter.kt
package dev.anvil

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager

/**
 * App lifecycle and device status.
 *
 * The core needs to know about backgrounding because Android restricts what a
 * backgrounded app may do with radios and the microphone — and because a
 * backgrounded device is a poor relay candidate even when its network looks
 * excellent.
 *
 * Battery and charging state feed relay election (§37): a plugged-in phone is
 * the right device to carry a room, and one at 8% is not.
 *
 * PHASE1.
 */
class LifecycleAdapter(
    private val context: Context,
    private val emit: (PlatformEvent) -> Unit,
) {

    private var foreground = false
    private var batteryReceiver: BroadcastReceiver? = null

    fun start() {
        if (batteryReceiver == null) {
            val receiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    emitStatus(intent)
                }
            }
            batteryReceiver = receiver
            context.registerReceiver(receiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        }
        foreground()
    }

    fun stop() {
        batteryReceiver?.let { receiver ->
            context.unregisterReceiver(receiver)
            batteryReceiver = null
        }
        background()
    }

    fun foreground() {
        if (!foreground) {
            foreground = true
            emit(PlatformEvent.LifecycleChanged(true))
        }
    }

    fun background() {
        if (foreground) {
            foreground = false
            emit(PlatformEvent.LifecycleChanged(false))
        }
    }

    private fun emitStatus(intent: Intent) {
        val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
        val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
        val percentage = if (level >= 0 && scale > 0) (level * 100 / scale).coerceIn(0, 100) else null
        val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
        val charging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
            status == BatteryManager.BATTERY_STATUS_FULL
        emit(PlatformEvent.DeviceStatus(percentage, charging, false))
    }
}

File: apps/mobile/ios/Runner/LifecycleAdapter.swift
import UIKit

/// App lifecycle and device status.
///
/// Foreground/background matters because iOS restricts radio and microphone
/// access in the background, and because a backgrounded device makes a poor
/// relay however good its network looks.
///
/// Battery and charging state feed relay election (§37). Note
/// `UIDevice.current.isBatteryMonitoringEnabled` must be set before battery
/// level reads anything but -1 — a small thing that silently produces a
/// nonsense relay score if forgotten.
///
/// PHASE1.
final class LifecycleAdapter {

    private let emit: (PlatformEvent) -> Void
    private var observers: [NSObjectProtocol] = []

    init(emit: @escaping (PlatformEvent) -> Void) {
        self.emit = emit
    }

    func start() {
        guard observers.isEmpty else { return }
        UIDevice.current.isBatteryMonitoringEnabled = true
        let center = NotificationCenter.default
        observers.append(center.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in self?.emit(.lifecycleChanged(foreground: false)) })
        observers.append(center.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in self?.emit(.lifecycleChanged(foreground: true)) })
        observers.append(center.addObserver(
            forName: UIDevice.batteryLevelDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in self?.emitStatus() })
        observers.append(center.addObserver(
            forName: UIDevice.batteryStateDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in self?.emitStatus() })
        if #available(iOS 11.0, *) {
            observers.append(center.addObserver(
                forName: ProcessInfo.thermalStateDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in self?.emitStatus() })
        }
        emit(.lifecycleChanged(foreground: true))
        emitStatus()
    }

    func stop() {
        observers.forEach(NotificationCenter.default.removeObserver)
        observers.removeAll()
        UIDevice.current.isBatteryMonitoringEnabled = false
    }

    private func emitStatus() {
        let level = UIDevice.current.batteryLevel
        let percentage: Int? = level >= 0 ? Int((level * 100).rounded()) : nil
        let state = UIDevice.current.batteryState
        let charging = state == .charging || state == .full
        let throttled: Bool
        if #available(iOS 11.0, *) {
            throttled = ProcessInfo.processInfo.thermalState == .serious ||
                ProcessInfo.processInfo.thermalState == .critical
        } else {
            throttled = false
        }
        emit(.deviceStatus(
            batteryPct: percentage,
            charging: charging,
            thermallyThrottled: throttled
        ))
    }
}
